### ISSUE TITLE
css selectors doesn't work?

### ISSUE THREAD

**Issue by danielzzz** [tebelorg/TA.Gui#12]
hi, I am testing it on the https://google.es page:
```
http://google.es
show center div
```
but it returns: ERROR - cannot find center div

is it a bug? or am I doing something wrong?

In general nothing with hierarchy worked for me:
```
center div
center>div
```
adding style in the browser (chrome):
```
center div {
    color: red;
}
```
correctly detects elements and change their color.

on another page it worked only when selected using class:
```
<div class="some_class another_class"><label>test</label></div>
```
.some_class.another_class label **WORKED** (this came from the chrome extension)
div.some_class label **DIDN'T WORK**
div label **DIDN'T WORK**
div>label **DIDN'T WORK**

cheers, dan

**Comment by danielzzz** [comment]
it may be related, I have the following page:

```
<!DOCTYPE html>
<html lang="">
  <head>
    
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta charset="utf-8"/>    
    <title>
        Admin/Users        
    </title>
    <link href="/favicon.ico" type="image/x-icon" rel="icon"/><link href="/favicon.ico" type="image/x-icon" rel="shortcut icon"/>
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.2/css/bootstrap.min.css"/>    <!-- <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css"> -->
    <link rel="stylesheet" href="/css/style.css"/>    <!--
      <link rel="stylesheet" href="/css/base.css"/>      <link rel="stylesheet" href="/css/cake.css"/>    -->

    <!-- Custom Fonts -->
    <link href="http://fonts.googleapis.com/css?family=Montserrat:400,700" rel="stylesheet" type="text/css">
    <link href="http://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic" rel="stylesheet" type="text/css">

    <!-- Bootstrap Core CSS - Uses Bootswatch Flatly Theme: http://bootswatch.com/flatly/ -->
    <link href="/css/bootstrap.css" rel="stylesheet">
    <link href="/css/font-awesome.min.css" rel="stylesheet">
    <link href="/css/admin.css" rel="stylesheet">

            
    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
    <style type="text/css">
      body {
        padding-top: 40px;
        padding-bottom: 40px;
        background-color: #eee;
      }

      .form-signin {
        max-width: 330px;
        padding: 15px;
        margin: 0 auto;
      }

    </style>
  </head>
  <body>
    


        <div id="flashMessage">
                        
        </div>

        <div class="form-signin">
          
            
            
<!--
<div class="grid_4 alpha omega prefix_4 suffix_4">

    <form id="loginForm" action="/admin/users/login/" method="post">

        <div class="widget-title">
            <span class="icon"><i class="icon-th-2"></i></span>
            <h5>ADMINISTRATION AREA</h5>
        </div>

        <fieldset>

            <div class="grid_3">
            <div class="input text required"><label for="UserUsername">Username</label><input name="data[User][username]" type="text" maxlength="50" id="UserUsername" /></div>            </div>

            <hr class="form" />

            <div class="grid_3">
            <div class="input password"><label for="UserUserpassword">Password</label><input type="password" name="data[User][userpassword]" id="UserUserpassword" /></div>            </div>

            <hr class="form" />

            <div class="grid_3">
                <a href="/users/recoverpassword/">I forgot my password</a>            </div>

        </fieldset>

        <hr />

        <button class="default_button right" type="submit">Login</button>
        <button class="default_button" onclick="window.location.href='/';return false;">Cancel</button>

    </form>

</div>

-->
<form method="post" accept-charset="utf-8" role="form" action="/admin/users/login"><div style="display:none;"><input type="hidden" name="_method" value="POST"/></div><fieldset>
    <div class="widget-title">
            <span class="icon"><i class="icon-th-2"></i></span>
            <h5>ADMINISTRATION AREA</h5>
        </div>
    <div class="grid_3">
    <div class="form-group text"><label class="control-label" for="username">Login</label><input type="text" name="username" id="username" class="form-control"/></div>    <div class="form-group password"><label class="control-label" for="userpassword">Pasword</label><input type="password" name="userpassword" id="userpassword" class="form-control"/></div>    </div>
</fieldset>
<button type="submit" class="btn btn-default">Login</button> 
<button type="button" onclick="window.location.href='/';return false;" class="btn btn-default">Cancel</button></form></div>
         
        </div>
        
        <hr>
        <div class="container">
          

          <footer class="text-center center">
            <p>&copy; 2016</p>
          </footer>
        </div>
          
        
    
    
    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
    <!-- Include all compiled plugins (below), or include individual files as needed -->
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.2/js/bootstrap.min.js"></script>
      <script id="__debug_kit" data-id="b2b1ee16-a894-454b-82e6-41b24dcedb7c" data-url="http://rutas.fontoasis.es/" src="/debug_kit/js/toolbar.js"></script></body>
</html>
```

and the following script:
```
http://[page url]
show body
```

displays the content of the style tag instead:
```
 body {
        padding-top: 40px;
        padding-bottom: 40px;
        background-color: #eee;
      }

      .form-signin {
        max-width: 330px;
        padding: 15px;
        margin: 0 auto;
      }
```

**Comment by danielzzz** [comment]
also a related question: does tagui handles session cookies? I am not sure it passes through the login form. debug shows a bit confusing info.

**Comment by kensoh** [comment]
Hi @danielzzz, please try if center div works now. What happens is I am attempting to let users 'dynamically' use css + xpath (full xpath or some string within id, name, class, title, text()) of the element).

The pro is in doing so, most commonly used cases could be expressed (with full xpath as most accurate) without users specifying whether the element identifier is css, xpath component name or full xpath. However, the catch is some css or elements will not be interpreted 1-to-1.

Inside the file /src/tagui_header.js is the logic to auto-select provided identifier in this order - full-xpath, full-css, id, name, class, title, text(). I use below function to decide whether something is a css selector. I've just commited 'center ' in and show center div works now.

Would you be able to take a quick look to recommend other commonly used css selector? If make it too comprehensive, the catch is the logic will overly favour css instead of xpath and have more misinterpretations of xpath selectors. I compiled these by asking around and doing some research, as I'm more familiar with using xpath selectors.

```js
// checking if selector is css selector
function is_css_selector(selector) {
if (selector.length == 0) return false; if (selector.charAt(0) == '[') return true;
if (selector.charAt(0) == '.') return true; if (selector.charAt(0) == '#') return true;
if (selector.substring(0,4) == 'div ') return true; if (selector.substring(0,4) == 'div:') return true;
if (selector.substring(0,4) == 'div[') return true; if (selector.substring(0,2) == 'p ') return true;
if (selector.substring(0,2) == 'p:') return true; if (selector.substring(0,2) == 'p[') return true;
if (selector.substring(0,6) == 'input ') return true; if (selector.substring(0,6) == 'input:') return true;
if (selector.substring(0,6) == 'input[') return true; if (selector.substring(0,7) == 'button ') return true;
if (selector.substring(0,7) == 'button:') return true; if (selector.substring(0,7) == 'button[') return true;
if (selector.substring(0,3) == 'li ') return true; if (selector.substring(0,3) == 'li:') return true;
if (selector.substring(0,3) == 'li[') return true; if (selector.substring(0,3) == 'tr ') return true;
if (selector.substring(0,3) == 'tr:') return true; if (selector.substring(0,3) == 'tr[') return true;
if (selector.substring(0,5) == 'span ') return true; if (selector.substring(0,5) == 'span:') return true;
if (selector.substring(0,5) == 'span[') return true;
if (selector.substring(0,3) == 'h1 ') return true; if (selector.substring(0,3) == 'h1:') return true;
if (selector.substring(0,3) == 'h1[') return true; if (selector.substring(0,3) == 'h2 ') return true;
if (selector.substring(0,3) == 'h2:') return true; if (selector.substring(0,3) == 'h2[') return true;
if (selector.substring(0,3) == 'h3 ') return true; if (selector.substring(0,3) == 'h3:') return true;
if (selector.substring(0,3) == 'h3[') return true; if (selector.substring(0,3) == 'h4 ') return true;
if (selector.substring(0,3) == 'h4:') return true; if (selector.substring(0,3) == 'h4[') return true;
if (selector.substring(0,3) == 'h5 ') return true; if (selector.substring(0,3) == 'h5:') return true;
if (selector.substring(0,3) == 'h5[') return true; if (selector.substring(0,3) == 'h6 ') return true;
if (selector.substring(0,3) == 'h6:') return true; if (selector.substring(0,3) == 'h6[') return true;
if (selector.substring(0,7) == 'center ') return true;
return false;}
```

**Comment by danielzzz** [comment]
maybe a solutions would be to use some indicator for xpath and other stuff
something like:
checkXpath
checkCss
along with guess all "check"
this might clear some confusion.

this last commit fixed "center div" for me but not "center>div" for example.
thanks!

**Comment by kensoh** [comment]
Yes cookies and logging in should be manageable, and already supported by underlying PhantomJS headless browser. Yes running with debug shows the actual raw feeds from PhantomJS and contains the whole chunk of raw verbose outputs from PhantomJS.

Are you running it on macOS? Would you be able to try SlimerJS + Firefox and running using firefox option? Running the automation in visible mode should normally be an easier way to test while creating the flow.

**Comment by danielzzz** [comment]
no, I have linux here. tried to run it in the firefox mode but it didn't work

**Comment by kensoh** [comment]
Yep, I didn't commit for 'center>', it passes through the checkCss step and goes into the general step. Will you be able to help take a look at the is_css_selector function above to see if what is checked there is valid? It sounds like besides using ':', '[' right after the initial keyword, '>' is also a commonly used in css selector language.

I see.. Can you tell me more about what error messages you see with firefox option? For Linux without a GUI it won't work as it requires a visible Firefox browser installed and SlimerJS installed. For Linux with GUI it should be expected to work. A common cause is [this](https://docs.slimerjs.org/current/installation.html#using-unstable-version-or-very-old-versions-of-firefox) version setting for security reason that breaks SlimerJS/Firefox integration.

**Comment by danielzzz** [comment]
@kensoh instead of trying to nail down all possible cases maybe this would be a better solution to check if an expression is a valid css selector:
http://stackoverflow.com/questions/34849001/check-if-css-selector-is-valid

**Comment by kensoh** [comment]
Thanks @danielzzz ! That is worth a try, by design I want it to pro-xpath, checking if css selector exists may cause some element descriptions for id/name/text() etc to not be able to reach xpath.

In any case, I can try shifting the orders of checks and as a last resort change the syntax so that users have to explicitly declare whether the identifier they are providing is meant for xpath or css.

Kind Regards,
Ken

**Comment by danielzzz** [comment]
thanks!

**Comment by kensoh** [comment]
I've committed the fix for this issue with following comments. what you've suggested might turns out to be the right balance between favoring xpath or css more. Thanks @danielzzz for taking time to poke on gaps on TA.Gui and make it better!

Let me know if using the new code still is not working for you when using css selectors, so that I can reopen the issue to work together on this with you.

Kind Regards,
Ken

experimental change for #12, skip css syntax checks totally. if this turns out to be a good balance for having dynamic xpath/css combined into a single selector, then the function is_css_selector can be removed totally from code in tagui_header.js

did a few sanity tests and looks good! thanks to @danielz and @dilab for your inputs on css selectors. i'm pro xpath and make this dynamic selector pro-xpath by design. this change tilts the bias slightly to be more balanced. will look out for user complaints and potential issues arising from my own using of this new logic.

**Comment by danielzzz** [comment]
thanks!
